src/main/java/hudson/plugins/s3/callable/S3GzipCallable.java

```diff
@@ -21,23 +24,47 @@ public S3GzipCallable(String accessKey, Secret secretKey, boolean useRole, Desti
     @Override
     public String invoke(FilePath file) throws IOException, InterruptedException {
         final File localFile = File.createTempFile("s3plugin", ".bin");
+        Upload upload = null;
 
-        try (InputStream inputStream = file.read()) {
-            try (OutputStream outputStream = new FileOutputStream(localFile)) {
-                try (OutputStream gzipStream = new GZIPOutputStream(outputStream, true)) {
-                    IOUtils.copy(inputStream, gzipStream);
-                    gzipStream.flush();
+        try {
+            try (InputStream inputStream = file.read()) {
+                try (OutputStream outputStream = new FileOutputStream(localFile)) {
+                    try (OutputStream gzipStream = new GZIPOutputStream(outputStream, true)) {
+                        IOUtils.copy(inputStream, gzipStream);
+                        gzipStream.flush();
+                    }
                 }
             }
-        }
 
-        final InputStream gzipedStream = new FileInputStream(localFile);
-        final ObjectMetadata metadata = buildMetadata(file);
-        metadata.setContentEncoding("gzip");
-        metadata.setContentLength(localFile.length());
+            final InputStream gzipedStream = new FileInputStream(localFile);
+            final ObjectMetadata metadata = buildMetadata(file);
+            metadata.setContentEncoding("gzip");
+            metadata.setContentLength(localFile.length());
+
+            upload = Uploads.getInstance().startUploading(getTransferManager(), file, gzipedStream, getDest().bucketName, getDest().objectName, metadata);
 
-        Uploads.getInstance().startUploading(getTransferManager(), file, gzipedStream, getDest().bucketName, getDest().objectName, metadata);
+            String md5 = MD5.generateFromFile(localFile);
 
-        return MD5.generateFromFile(localFile);
+            upload.addProgressListener(new ProgressListener() {
+                    @Override
+                    public void progressChanged(ProgressEvent event) {
+                        switch (event.getEventType()) {
+                        case TRANSFER_CANCELED_EVENT:
```

Comment:
This seems to be an independent bug, not directly related to the gzipped temporary files. I'm not quite familiar enough with the S3 plugin to deal with that, and anyway it's probably less confusing if the cleanup bug gets fixed separately, not in this PR.